core/src/main/java/org/infinispan/persistence/async/AdvancedAsyncCacheLoader.java

```diff
@@ -1,120 +1,84 @@
-               @Override
-               public void processEntry(MarshalledEntry marshalledEntry, TaskContext taskContext) throws InterruptedException {
-                  result.add(marshalledEntry.getKey());
-               }
-            }, executor, false, false);
-         }
+   @Override
+   public Publisher<K> publishKeys(Predicate<? super K> filter) {
+      State state = this.state.get();
+      ByRef<Boolean> hadClear = new ByRef<>(Boolean.FALSE);
+      Map<Object, Modification> modificationMap = state.flattenModifications(hadClear);
+      if (modificationMap.isEmpty()) {
+         return advancedLoader().publishKeys(filter);
       }
 
-      // merge keys of the current State
-      for (Modification mod : s.modifications.values()) {
-         switch (mod.getType()) {
-            case STORE:
-               Object key = ((Store) mod).getStoredValue().getKey();
-               if (filter == null || filter.accept(key))
-                  result.add(key);
-               break;
-            case REMOVE:
-               result.remove(((Remove) mod).getKey());
-               break;
-         }
+      Publisher<K> modPublisher = Flowable.fromIterable(modificationMap.entrySet())
+            // REMOVE we ignore, LIST and CLEAR aren't possible
+            .filter(me -> Modification.Type.STORE == me.getValue().getType())
+            .map(e -> (K) e.getKey());
+
+      if (hadClear.get() == Boolean.TRUE) {
+         return modPublisher;
+      }
+      if (filter == null) {
+         filter = k -> !modificationMap.containsKey(k);
+      } else {
+         filter = filter.and(k -> !modificationMap.containsKey(k));
       }
+      return Flowable.merge(modPublisher, advancedLoader().publishKeys(filter));
    }
 
-
-   @SuppressWarnings("unchecked")
    @Override
-   public void process(KeyFilter keyFilter, CacheLoaderTask cacheLoaderTask, Executor executor, boolean loadValues, boolean loadMetadata) {
-
-      int batchSize = 100;
-      ExecutorAllCompletionService eacs = new ExecutorAllCompletionService(executor);
-      final TaskContext taskContext = new TaskContextImpl();
-
-      Set<Object> allKeys = new ConcurrentHashSet<>();
-      Set<Object> batch = new HashSet<Object>();
-      loadAllKeys(state.get(), allKeys, keyFilter, executor);
-      for (Iterator it = allKeys.iterator(); it.hasNext(); ) {
-         batch.add(it.next());
-         if (batch.size() == batchSize) {
-            final Set<Object> toHandle = batch;
-            batch = new HashSet<>(batchSize);
-            submitProcessTask(cacheLoaderTask, eacs, taskContext, toHandle);
-         }
+   public Publisher<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
+      State state = this.state.get();
+      ByRef<Boolean> hadClear = new ByRef<>(Boolean.FALSE);
+      Map<Object, Modification> modificationMap = state.flattenModifications(hadClear);
```

Comment:
Unfortunately we need to make a copy, otherwise the state can be updated after iterating through it and causing us to exclude an entry from the store that we shouldn't have.

Reply:
Yep my thoughts as well.